File: native/ios/FireballNative/Core/FireballAPIClient.swift
import Foundation

private final class InvidiousLoginRedirectDelegate: NSObject, URLSessionTaskDelegate, @unchecked Sendable {
    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
    }
}

final class FireballAPIClient {
    private let session: URLSession
    private let loginDelegate = InvidiousLoginRedirectDelegate()
    private lazy var loginSession: URLSession = {
        let cfg = URLSessionConfiguration.ephemeral
        cfg.httpShouldSetCookies = false
        cfg.httpCookieStorage = nil
        cfg.httpCookieAcceptPolicy = .never
        return URLSession(configuration: cfg, delegate: loginDelegate, delegateQueue: nil)
    }()

    static let defaultUserAgent =
        "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148 FireballNative/1.0"

    init(session: URLSession = .shared) {
        self.session = session
    }

    private func applyDefaultHeaders(_ request: inout URLRequest) {
        request.setValue(Self.defaultUserAgent, forHTTPHeaderField: "User-Agent")
    }

    private func trimTrailingSlashes(_ url: String) -> String {
        var s = url
        while s.last == "/" { s.removeLast() }
        return s
    }

    func iTunesSearch(term: String, limit: Int = 25) async throws -> Data {
        var components = URLComponents(string: "https://itunes.apple.com/search")!
        components.queryItems = [
            URLQueryItem(name: "media", value: "music"),
            URLQueryItem(name: "entity", value: "song"),
            URLQueryItem(name: "term", value: term),
            URLQueryItem(name: "limit", value: "\(limit)"),
        ]
        var request = URLRequest(url: components.url!)
        applyDefaultHeaders(&request)
        let (data, _) = try await session.data(for: request)
        return data
    }

    func invidiousSearch(instanceURL: String, query: String, sid: String? = nil) async throws -> Data {
        let base = trimTrailingSlashes(instanceURL)
        var c = URLComponents(string: "\(base)/api/v1/search")!
        c.queryItems = [
            URLQueryItem(name: "q", value: query),
            URLQueryItem(name: "type", value: "video"),
        ]
        guard let url = c.url else { throw URLError(.badURL) }
        var request = URLRequest(url: url)
        applyDefaultHeaders(&request)
        if let sid, !sid.isEmpty {
            request.setValue("SID=\(sid)", forHTTPHeaderField: "Cookie")
        }
        let (data, _) = try await session.data(for: request)
        return data
    }

    /// `local=true` proxies streams through the instance (recommended for playback).
    /// `local=true` asks Invidious to proxy stream URLs (recommended for playback).
    func invidiousVideo(
        instanceURL: String,
        videoId: String,
        sid: String? = nil,
        local: Bool = true
    ) async throws -> Data {
        let base = trimTrailingSlashes(instanceURL)
        let encodedId = videoId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? videoId
        var components = URLComponents(string: "\(base)/api/v1/videos/\(encodedId)")!
        if local {
            components.queryItems = [URLQueryItem(name: "local", value: "true")]
        }
        guard let url = components.url else { throw URLError(.badURL) }
        var request = URLRequest(url: url)
        applyDefaultHeaders(&request)
        if let sid, !sid.isEmpty {
            request.setValue("SID=\(sid)", forHTTPHeaderField: "Cookie")
        }
        let (data, _) = try await session.data(for: request)
        return data
    }

    /// POST to Invidious login; does not follow redirects so `Set-Cookie` / `SID` can be read from the response.
    func invidiousLogin(instanceURL: String, username: String, password: String) async throws -> (sid: String, username: String) {
        let base = trimTrailingSlashes(instanceURL)
        guard let url = URL(string: "\(base)/login?type=invidious") else { throw URLError(.badURL) }

        func tryLogin(body: String) async throws -> (sid: String, code: Int) {
            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
            request.setValue("\(base)/login", forHTTPHeaderField: "Referer")
            request.setValue(base, forHTTPHeaderField: "Origin")
            applyDefaultHeaders(&request)
            request.httpBody = body.data(using: .utf8)
            let (_, response) = try await loginSession.data(for: request)
            let code = (response as? HTTPURLResponse)?.statusCode ?? -1
            let sid = extractSid(from: response) ?? ""
            return (sid, code)
        }

        let encUser = username.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? username
        let encPass = password.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? password
        var (sid, code) = try await tryLogin(body: "email=\(encUser)&password=\(encPass)")
        if sid.isEmpty {
            (sid, code) = try await tryLogin(body: "email=\(encUser)&password=\(encPass)&action=signin")
        }
        if sid.isEmpty {
            let reason = (code == 401 || code == 403) ? "wrong credentials" : "status=\(code), captcha or login disabled"
            throw NSError(domain: "InvidiousLogin", code: code, userInfo: [NSLocalizedDescriptionKey: "Invidious login failed (\(reason))."])
        }
        return (sid: sid, username: username)
    }

    private func extractSid(from response: URLResponse) -> String? {
        guard let http = response as? HTTPURLResponse, let url = http.url else { return nil }
        var headerDict = [String: String]()
        for (key, value) in http.allHeaderFields {
            if let k = key as? String, let v = value as? String {
                headerDict[k] = v
            }
        }
        let cookies = HTTPCookie.cookies(withResponseHeaderFields: headerDict, for: url)
        if let sid = cookies.first(where: { $0.name.caseInsensitiveCompare("SID") == .orderedSame })?.value, !sid.isEmpty {
            return sid
        }
        var combined = ""
        for (k, v) in headerDict where k.caseInsensitiveCompare("Set-Cookie") == .orderedSame {
            combined += v + "\n"
        }
        if let r = try? NSRegularExpression(pattern: "SID=([^;]+)", options: .caseInsensitive),
           let m = r.firstMatch(in: combined, range: NSRange(combined.startIndex..., in: combined)),
           m.numberOfRanges > 1,
           let range = Range(m.range(at: 1), in: combined) {
            return String(combined[range]).trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return nil
    }

    func invidiousPlaylistDetail(instanceURL: String, sid: String?, playlistId: String) async throws -> Data {
        let base = trimTrailingSlashes(instanceURL)
        let enc = playlistId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? playlistId
        guard let url = URL(string: "\(base)/api/v1/playlists/\(enc)") else { throw URLError(.badURL) }
        var request = URLRequest(url: url)
        applyDefaultHeaders(&request)
        if let sid, !sid.isEmpty {
            request.setValue("SID=\(sid)", forHTTPHeaderField: "Cookie")
        }
        let (data, _) = try await session.data(for: request)
        return data
    }

    func invidiousAuthPlaylists(instanceURL: String, sid: String) async throws -> Data {
        let base = trimTrailingSlashes(instanceURL)
        guard let url = URL(string: "\(base)/api/v1/auth/playlists") else { throw URLError(.badURL) }
        var request = URLRequest(url: url)
        applyDefaultHeaders(&request)
        request.setValue("SID=\(sid)", forHTTPHeaderField: "Cookie")
        let (data, _) = try await session.data(for: request)
        return data
    }

    func invidiousCreatePlaylist(instanceURL: String, sid: String, title: String, privacy: String) async throws -> String {
        let base = trimTrailingSlashes(instanceURL)
        guard let url = URL(string: "\(base)/api/v1/auth/playlists") else { throw URLError(.badURL) }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        applyDefaultHeaders(&request)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("SID=\(sid)", forHTTPHeaderField: "Cookie")
        let body: [String: Any] = ["title": title, "privacy": privacy]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, _) = try await session.data(for: request)
        let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        return (obj?["playlistId"] as? String) ?? ""
    }

    func invidiousAddVideoToPlaylist(instanceURL: String, sid: String, playlistId: String, videoId: String) async throws {
        let base = trimTrailingSlashes(instanceURL)
        let encPl = playlistId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? playlistId
        guard let url = URL(string: "\(base)/api/v1/auth/playlists/\(encPl)/videos") else { throw URLError(.badURL) }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        applyDefaultHeaders(&request)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("SID=\(sid)", forHTTPHeaderField: "Cookie")
        request.httpBody = try JSONSerialization.data(withJSONObject: ["videoId": videoId])
        _ = try await session.data(for: request)
    }

    func lrclibSearch(track: String, artist: String) async throws -> Data {
        var components = URLComponents(string: "https://lrclib.net/api/search")!
        components.queryItems = [
            URLQueryItem(name: "track_name", value: track),
            URLQueryItem(name: "artist_name", value: artist),
        ]
        var request = URLRequest(url: components.url!)
        applyDefaultHeaders(&request)
        let (data, _) = try await session.data(for: request)
        return data
    }

    func neteaseSearch(query: String) async throws -> Data {
        var components = URLComponents(string: "https://music.163.com/api/search/get")!
        components.queryItems = [
            URLQueryItem(name: "s", value: query),
            URLQueryItem(name: "type", value: "1"),
            URLQueryItem(name: "limit", value: "10"),
        ]
        var request = URLRequest(url: components.url!)
        applyDefaultHeaders(&request)
        let (data, _) = try await session.data(for: request)
        return data
    }

    func neteaseLyrics(songId: String) async throws -> Data {
        var components = URLComponents(string: "https://music.163.com/api/song/lyric")!
        components.queryItems = [
            URLQueryItem(name: "id", value: songId),
            URLQueryItem(name: "lv", value: "1"),
            URLQueryItem(name: "kv", value: "1"),
            URLQueryItem(name: "tv", value: "-1"),
        ]
        var request = URLRequest(url: components.url!)
        applyDefaultHeaders(&request)
        let (data, _) = try await session.data(for: request)
        return data
    }

    func ollamaGenerate(ollamaUrl: String, model: String, prompt: String) async throws -> Data {
        let base = ollamaUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        var request = URLRequest(url: URL(string: "\(base)/api/chat")!)
        request.httpMethod = "POST"
        applyDefaultHeaders(&request)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "model": model,
            "stream": false,
            "messages": [["role": "user", "content": prompt]],
        ])
        return try await session.data(for: request).0
    }
}

File: native/android/core-data/src/main/kotlin/com/fireball/nativeapp/core/data/FireballApiClient.kt
package com.fireball.nativeapp.core.data

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.header
import io.ktor.client.request.get
import io.ktor.client.request.post
import io.ktor.client.request.parameter
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.contentType
import io.ktor.http.encodeURLParameter
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import java.net.HttpURLConnection
import java.net.URL

class FireballApiClient(
    private val httpClient: HttpClient,
) {
    suspend fun healthyInvidiousInstances(): List<String> =
        InvidiousInstanceProvider.fetchHealthyInstances(httpClient)
    /** iTunes returns JSON with Content-Type often `application/javascript`; avoid Ktor JSON negotiation missing it. */
    private val looseJson = Json { ignoreUnknownKeys = true; isLenient = true }

    suspend fun itunesSearch(term: String, limit: Int = 25): JsonObject {
        val bodyText = httpClient.get("https://itunes.apple.com/search") {
            parameter("media", "music")
            parameter("entity", "song")
            parameter("term", term)
            parameter("limit", limit)
            parameter("country", "US")
        }.bodyAsText()
        return looseJson.parseToJsonElement(bodyText).jsonObject
    }

    suspend fun invidiousSearch(instanceUrl: String, query: String, sid: String? = null): JsonArray {
        val base = instanceUrl.trimEnd('/')
        return httpClient.get("$base/api/v1/search") {
            parameter("q", query)
            parameter("type", "video")
            if (!sid.isNullOrBlank()) header(HttpHeaders.Cookie, "SID=$sid")
        }.body()
    }

    /**
     * @param local When true, Invidious proxies stream URLs through the instance (recommended for mobile playback).
     * See https://docs.invidious.io/url-parameters/
     */
    suspend fun invidiousVideo(
        instanceUrl: String,
        videoId: String,
        sid: String? = null,
        local: Boolean = true,
    ): JsonObject {
        val base = instanceUrl.trimEnd('/')
        return httpClient.get("$base/api/v1/videos/$videoId") {
            if (local) parameter("local", "true")
            if (!sid.isNullOrBlank()) header(HttpHeaders.Cookie, "SID=$sid")
        }.body()
    }

    /** Lightweight health probe — https://docs.invidious.io/api/#get-apiv1stats */
    suspend fun invidiousPing(instanceUrl: String): Boolean = runCatching {
        val base = instanceUrl.trimEnd('/')
        httpClient.get("$base/api/v1/stats").bodyAsText()
        true
    }.getOrDefault(false)

    suspend fun pipedSearch(instanceUrl: String, query: String): JsonArray {
        val base = instanceUrl.trimEnd('/')
        val response = httpClient.get("$base/search") {
            parameter("q", query)
            parameter("filter", "music_songs")
        }.body<JsonObject>()
        return response["items"]?.jsonArray ?: JsonArray(emptyList())
    }

    suspend fun pipedStream(instanceUrl: String, videoId: String): JsonObject {
        val base = instanceUrl.trimEnd('/')
        return httpClient.get("$base/streams/$videoId").body()
    }

    suspend fun invidiousLogin(instanceUrl: String, username: String, password: String): Pair<String, String> = kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
        val base = instanceUrl.trimEnd('/')
        val loginUrl = URL("$base/login?type=invidious")

        fun tryLogin(body: String): Pair<String, Int> {
            val conn = (loginUrl.openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                instanceFollowRedirects = false // Keep 302 response to inspect Set-Cookie SID
                connectTimeout = 15_000
                readTimeout = 15_000
                doOutput = true
                setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
                setRequestProperty("Referer", "$base/login")
                setRequestProperty("Origin", base)
                setRequestProperty("User-Agent", "Mozilla/5.0 (compatible; Fireball/1.0; +https://github.com/fireball)")
                outputStream.use { os -> os.write(body.toByteArray()) }
            }

            val code = conn.responseCode
            val setCookie = buildList {
                conn.headerFields?.forEach { (key, values) ->
                    if (key != null && key.equals("Set-Cookie", ignoreCase = true)) {
                        addAll(values.orEmpty())
                    }
                }
            }
            val sid = setCookie
                .firstNotNullOfOrNull { Regex("""SID=([^;]+)""", RegexOption.IGNORE_CASE).find(it)?.groupValues?.getOrNull(1) }
                ?.trim()
                .orEmpty()
            return sid to code
        }

        var body = "email=${username.encodeURLParameter()}&password=${password.encodeURLParameter()}"
        var (sid, code) = tryLogin(body)

        if (sid.isBlank()) {
            body = "$body&action=signin"
            val retry = tryLogin(body)
            sid = retry.first
            code = retry.second
        }

        if (sid.isBlank()) {
            val reason = when (code) {
                401, 403 -> "wrong credentials"
                else -> "status=$code, captcha required or login disabled"
            }
            error("Invidious login failed ($reason).")
        }
        return@withContext sid to username
    }

    suspend fun invidiousPlaylistDetail(instanceUrl: String, sid: String?, playlistId: String): JsonObject {
        val base = instanceUrl.trimEnd('/')
        return httpClient.get("$base/api/v1/playlists/$playlistId") {
            if (!sid.isNullOrBlank()) header(HttpHeaders.Cookie, "SID=$sid")
        }.body()
    }

    suspend fun invidiousAuthPlaylists(instanceUrl: String, sid: String): JsonArray {
        val base = instanceUrl.trimEnd('/')
        return httpClient.get("$base/api/v1/auth/playlists") {
            header(HttpHeaders.Cookie, "SID=$sid")
        }.body()
    }

    suspend fun invidiousCreatePlaylist(
        instanceUrl: String,
        sid: String,
        title: String,
        privacy: String
    ): String {
        val base = instanceUrl.trimEnd('/')
        val response = httpClient.post("$base/api/v1/auth/playlists") {
            header(HttpHeaders.Cookie, "SID=$sid")
            contentType(ContentType.Application.Json)
            setBody(buildJsonObject {
                put("title", title)
                put("privacy", privacy)
            })
        }.body<JsonObject>()
        return response["playlistId"]?.jsonPrimitive?.content.orEmpty()
    }

    suspend fun invidiousAddVideoToPlaylist(instanceUrl: String, sid: String, playlistId: String, videoId: String) {
        val base = instanceUrl.trimEnd('/')
        httpClient.post("$base/api/v1/auth/playlists/$playlistId/videos") {
            header(HttpHeaders.Cookie, "SID=$sid")
            contentType(ContentType.Application.Json)
            setBody(buildJsonObject { put("videoId", videoId) })
        }.bodyAsText()
    }

    suspend fun lrclibSearch(track: String, artist: String): JsonObject {
        return httpClient.get("https://lrclib.net/api/search") {
            parameter("track_name", track)
            parameter("artist_name", artist)
        }.body()
    }

    suspend fun neteaseSearch(query: String): JsonObject {
        return httpClient.get("https://music.163.com/api/search/get") {
            parameter("s", query)
            parameter("type", 1)
            parameter("limit", 10)
        }.body()
    }

    suspend fun neteaseLyrics(songId: String): JsonObject {
        return httpClient.get("https://music.163.com/api/song/lyric") {
            parameter("id", songId)
            parameter("lv", 1)
            parameter("kv", 1)
            parameter("tv", -1)
        }.body()
    }

    suspend fun ollamaGenerate(ollamaUrl: String, model: String, prompt: String): JsonObject {
        val base = ollamaUrl.trimEnd('/')
        return httpClient.post("$base/api/chat") {
            setBody(
                buildJsonObject {
                    put("model", model)
                    put("stream", false)
                    put("messages", kotlinx.serialization.json.buildJsonArray {
                        add(buildJsonObject {
                            put("role", "user")
                            put("content", prompt)
                        })
                    })
                }
            )
        }.body()
    }
}
